### ISSUE TITLE
CI: Add more compiler warnings

### ISSUE THREAD

**Issue by germasch** [ornladios/ADIOS2#1602]
@chuckatkins: Learning from the last minute MSVC troubles before the release, which were (in part) caused by variable length arrays on the stack, which is not a standard C++ feature, how about adding `-pedantic-errors` to the CI when the compiler is gcc/clang? This will catch this problem, and potentially avoid other non-standard use of C++.

It actually turns out that you'd have to add `-Wno-gnu-zero-variadic-macro-arguments` since gtest relies on that, and no fix appears to exist (yet):  https://github.com/google/googletest/issues/1419

A more limited approach would be to just avoid the use of variable length arrays with `-Werror=vla`.

**Comment by chuckatkins** [comment]
Another, possibly more robust option, would be to disable building with extensions.  The default mode for CMake when specifying language standards is to use the extended flavor, i.e. `-std=gnu++11` instead of `-std=c++11`.  We should be able to easily change that in the top level `CMakeLists.txt` for any compiler for which CMake understands (which is most).

**Comment by germasch** [comment]
I agree that enforcing standard conforming code (ie, C++11) would be good. I'm not sure whether one still has to add `-pedantic-errors`, anyway.

In practice, it might turn out that one can't really enforce everything strictly, like the gtest limitation I mentioned. But it's certainly worth trying ;)

**Comment by chuckatkins** [comment]
We can't add pedantic as it will break some of the third party code that we don't have control over but would should be able to enforce c++ standards modes.
